solace/src/main/java/org/apache/apex/malhar/solace/AbstractSolaceBaseInputOperator.java

```diff
@@ -0,0 +1,319 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.malhar.solace;
+
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import javax.validation.constraints.NotNull;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.solacesystems.jcsmp.BytesXMLMessage;
+import com.solacesystems.jcsmp.Consumer;
+import com.solacesystems.jcsmp.JCSMPChannelProperties;
+import com.solacesystems.jcsmp.JCSMPException;
+import com.solacesystems.jcsmp.JCSMPFactory;
+import com.solacesystems.jcsmp.JCSMPProperties;
+import com.solacesystems.jcsmp.JCSMPReconnectEventHandler;
+import com.solacesystems.jcsmp.JCSMPSession;
+import com.solacesystems.jcsmp.SessionEvent;
+import com.solacesystems.jcsmp.SessionEventArgs;
+import com.solacesystems.jcsmp.SessionEventHandler;
+import com.solacesystems.jcsmp.XMLMessageListener;
+import com.datatorrent.api.Context;
+import com.datatorrent.api.InputOperator;
+import com.datatorrent.api.Operator;
+import com.datatorrent.api.Operator.CheckpointListener;
+import com.datatorrent.common.util.BaseOperator;
+import com.datatorrent.lib.io.IdempotentStorageManager;
+import com.datatorrent.netlet.util.DTThrowable;
+
+@SuppressWarnings("unused")
+public abstract class AbstractSolaceBaseInputOperator<T> extends BaseOperator implements
+    InputOperator, Operator.ActivationListener<Context.OperatorContext>, CheckpointListener
+{
+
+  private static final Logger logger = LoggerFactory.getLogger(AbstractSolaceBaseInputOperator.class);
+
+  @NotNull
+  protected JCSMPProperties properties = new JCSMPProperties();
+  protected String connectRetries;
+  protected String reconnectRetries;
+  protected String unackedMessageLimit;
+
+
+  //protected IdempotentStorageManager idempotentStorageManager = new IdempotentStorageManager.FSIdempotentStorageManager();
+  protected FSOpsIdempotentStorageManager idempotentStorageManager = new FSOpsIdempotentStorageManager();
+
+  protected transient JCSMPFactory factory;
+  protected transient JCSMPSession session;
+
+  protected transient Consumer consumer;
+  protected transient Consumer reliableConsumer;
+
+  protected transient int operatorId;
+  protected long windowId;
+  protected transient long lastCompletedWId;
+
+  protected transient int emitCount;
+
+  protected transient volatile boolean DRFailover = false;
+  protected transient volatile boolean TCPDisconnected = false;
+
+  protected transient BlockingQueue<BytesXMLMessage> unackedMessages; // hosts the Solace messages that need to be acked when the streaming window is OK to remove
+  protected LinkedList<Long> inFlightMessageId = new LinkedList<Long>(); //keeps track of all in flight IDs since they are not necessarily sequential
+  protected transient ArrayBlockingQueue<BytesXMLMessage> arrivedTopicMessagesToProcess;
```

Comment:
This is declared transient. It will be good to describe (in a code comment) how you don't lose messages even if this is transient (e.g. similar to Kafka's EARLIEST or APPLICATION setting for the offset).